std-bits/base/src/main/java/org/enso/base/FileLineReader.java

```diff
@@ -0,0 +1,402 @@
+      throws IOException {
+    int size = rowMap.getSize();
+    if (index != -1 && size > index) {
+      return readLineByIndex(file, length, rowMap, index, charset);
+    }
+
+    // Start at the last known line and scan forward.
+    return forEachLine(file, length, rowMap, size - 1, index, charset, filter, null);
+  }
+
+  public static List<String> readLines(
+      File file,
+      long length,
+      LongArrayList rowMap,
+      int startAt,
+      int endAt,
+      Charset charset,
+      Function<ByteArrayOutputStreamWithContains, String> filter)
+      throws IOException {
+    List<String> result = new ArrayList<>();
+    forEachLine(
+        file, length, rowMap, startAt, endAt, charset, filter, (index, line) -> result.add(line));
+    return result;
+  }
+
+  // ** Scans forward in a file reading line by line.
+  // * @param file The file to read.
+  // * @param rowMap The rowMap to use.
+  // * @param startAt The index to start at.
+  // * @param endAt The index to end at (inclusive).
+  // * @param charset The charset to use.
+  // * @param filter The filter to apply to each line.
+  // * @param action The action to apply to each line (optional).
+  // * @return The last line read or null if end of file is reached.
+  // * *//
+  public static String forEachLine(
+      File file,
+      long length,
+      LongArrayList rowMap,
+      int startAt,
+      int endAt,
+      Charset charset,
+      Function<ByteArrayOutputStreamWithContains, String> filter,
+      BiConsumer<Integer, String> action)
+      throws IOException {
+    return innerForEachLine(
+        file, length, rowMap, startAt, endAt, charset, filter, action, new CancellationToken());
+  }
+
+  private static String innerForEachLine(
+      File file,
+      long length,
+      LongArrayList rowMap,
+      int startAt,
+      int endAt,
+      Charset charset,
+      Function<ByteArrayOutputStreamWithContains, String> filter,
+      BiConsumer<Integer, String> action,
+      CancellationToken cancellationToken)
+      throws IOException {
+    if (startAt >= rowMap.getSize()) {
+      throw new IndexOutOfBoundsException(startAt);
+    }
+    int index = action == null ? rowMap.getSize() - 1 : startAt;
+
+    long position = rowMap.get(index);
+    if (position >= length) {
+      return null;
+    }
+
+    boolean readAll = filter != null || action != null || endAt == -1;
+    var outputStream = new ByteArrayOutputStreamWithContains(128);
+    String output = null;
+
+    try (var stream = new FileInputStream(file)) {
+      var channel = stream.getChannel();
+
+      var bufferSize = (int) Math.min(BUFFER_SIZE, (length - position));
+      var truncated = bufferSize != (length - position);
+      var buffer = channel.map(FileChannel.MapMode.READ_ONLY, position, bufferSize);
+
+      // Loop until we either reach the required record or run out of data.
+      while (!cancellationToken.isCancelled
```

Comment:
It feels like this loop should have a safepoint, no?

Reply:
Line 276 runs a safepoint on each call.

Reply:
Ok